### PR TITLE
[MNT] improvements to modular CI framework - part 2, merge frameworks

### PR DESCRIPTION
**Pull request by fkiraly**
This PR makes multiple improvements to the modular CI framework (on top of #5713 and #5727).

High-level objectives:

* simplification: merge in one file, remove duplicative steps, preserve steps that are unique to new/old, consistent naming
* reduced runtime via arranging steps in conditional patterns, e.g., gating by simple pre-tests

Important: no reduced coverage is intended or caused by this PR.

Changes in detail:

* "new" and "old" framework are merged into a single `yml` file, resulting in a single action.
* the duplicative `code_quality` is removed, in favour of new, which seems to be using a more well-maintained file changes action
* "validate extras" aka "test install of soft dependency sets" is removed, as it is now duplicative because of #5727
* all testing is now gated behind the fast first-stage tests in `test-nosoftdeps`, as in the old framework
* naming of jobs is made uniform, e.g., start with "test", use of hyphens and underscores
*  the old CI jobs are still left in the PR, as we need to decide how to switch them off. There is also coverage of `pandas 1` which would be removed by naive delete, this should probably move to a CRON.

Depends on PR which should be merged first:
* https://github.com/sktime/sktime/pull/5713
* https://github.com/sktime/sktime/pull/5727

### CODE REVIEW

**Comment by fkiraly** [comment]
FYI @yarnabrina - looking at this, I think we should merge all "detect" steps into one, and the custom workflows into the main `yml`. What do you think?

**Comment by fkiraly** [comment]
> Can you please tell if you agree with both points, and if so how will these be tested? I am not clear how merging everything in one file will allow that, so need to think about it based on your goal.

1. the CRON job `test_all` is still there, so it will still run regularly as you inteded. I am just wondering, it does not run all estimators, does it? But this is as you wrote it originally.

2. the workflow triggered on PR is `test.yml`. The first job is indeed a check fo code quality on all files (meaning: linting and the like, not tests). If this passes, subsequent jobs are run, which cover your requirement "all affected objects must be tested on all supported environments with bare minimum additional dependencies as needed." As the jobs are copy-pastes of yours, I would think you would agree that this requirement is satisfied with the jobs in question.

**Comment by yarnabrina** [comment]
My question was in the context of this, which I intrpreted as there will be a single `test.yml` (or `main.yml`) file in `.github/workflows`.

> we should merge all "detect" steps into one, and the custom workflows into the main yml.

I am not able to follow how this can be then used to do both points as their filter, trigger, etc. conditions are different.

---

> I am just wondering, it does not run all estimators, does it?

That definitely was my goal. I tried to trigger all tests on all estimators, **_without_** sub-sampling (by using `--matrixdesign False`) and **_without_** differential testing (by using `--only_changed_modules False`), and this is done directly in the workflow file for "components-without-extras" (others in your terms) and through the action inputs for "components-with-extras" (module in your terms).

If something is not being tested, then that's unintended and a CI bug.

**Comment by fkiraly** [comment]
> My question was in the context of this, which I intrpreted as there will be a single `test.yml` (or `main.yml`) file in `.github/workflows`.

No, all I want is a single top-level file for workflows triggered by PR. I my desired end state:

* there can be other files triggered from the top-level file (but not by PR directly)
* there can be other workflows not trriggered by PR, but, e.g., by CRON

**Comment by fkiraly** [comment]
>> we should merge all "detect" steps into one, and the custom workflows into the main yml.

> I am not able to follow how this can be then used to do both points as their filter, trigger, etc. conditions are different.

We would trigger them all at the start and detect coditions at once.

**Comment by fkiraly** [comment]
> That definitely was my goal. I tried to trigger all tests on all estimators, **_without_** sub-sampling (by using `--matrixdesign False`) and **_without_** differential testing (by using `--only_changed_modules False`), and this is done directly in the workflow file for "components-without-extras" (others in your terms) and through the action inputs for "components-with-extras" (module in your terms).

Could you kindly point me to the line where the flag "test everything" is set? Would be appreciated.

**Comment by yarnabrina** [comment]
> Could you kindly point me to the line where the flag "test everything" is set? Would be appreciated.

For "components-without-extras", it is done in lines 141-142 in `.github/workflows/test_all.yml` workflow.

https://github.com/sktime/sktime/blob/b63f83a462297ab53541607cfe15a1602f0d577c/.github/workflows/test_all.yml#L141-L142

For "components-with-extras", it is passed to `.github/actions/test-component/action.yml` action in lines 89-90 in `.github/workflows/test_all.yml` workflow.

https://github.com/sktime/sktime/blob/b63f83a462297ab53541607cfe15a1602f0d577c/.github/workflows/test_all.yml#L89-L90

That then in turn uses them in lines 39-40.

https://github.com/sktime/sktime/blob/b63f83a462297ab53541607cfe15a1602f0d577c/.github/actions/test-component/action.yml#L39-L40

My expectation is that specifying `--matrixdesign False --only_changed_modules False` ensures to run unit tests for all (and not just modified) estimators in all (and not a subset) of supported environments for complete coverage.

**Comment by yarnabrina** [comment]
Can you please explain what does `test-nosoftdeps` job test? I see it calls `test_softdeps` target from `Makefile`, but it's very confusing name in my mind to know if it's using soft dependencies or not.

Is it "all" unit tests everywhere that should be able to run with just `sktime`? I have not used `backoff` and `httpx`, can you please tell where are these used for this testing?

**Comment by fkiraly** [comment]
> Can you please explain what does `test-nosoftdeps` job test? I see it calls `test_softdeps` target from `Makefile`, but it's very confusing name in my mind to know if it's using soft dependencies or not.

It tests that the package runs without soft dependencies installed, and that there are no imports of soft dependencies that would cause the package to break in a vanilla `pip install sktime` environment.

There are a couple of tests which certify from that, such as importing all modules (if it does not break, then there is no stray import), and constructing all estimators.

**Comment by yarnabrina** [comment]
@fkiraly I think the current PR summary (in the topmost comment above) describes #5713 + #5727 + #5785, and along with the conflict it's making me quite confused. Can you please tell what is the objective to this PR on its own? Is it dependent on #5727, or the reverse? I'll review in order to be more clear.

**Comment by fkiraly** [comment]
> Can you please tell what is the objective to this PR on its own?

Of course, I added the following paragraph at the top - hope its' clear?

High-level objectives:

* simplification: merge in one file, remove duplicative steps, preserve steps that are unique to new/old, consistent naming
* reduced runtime via arranging steps in conditional patterns, e.g., gating by simple pre-tests

Important: no reduced coverage is intended or caused by this PR.

> Is it dependent on https://github.com/sktime/sktime/pull/5727, or the reverse?

*This* PR depends on #5727, the logic in #5727 is necessary for former "code quality" being redundant with "test", and allowing to merge them into half the amount of jobs.

**Comment by fkiraly** [comment]
merged with changes from #5727, so ready for review

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -186,6 +189,47 @@ jobs:
...
+    steps:
+      - uses: dorny/paths-filter@v2
+        id: filter
+        with:
+          filters: |
+            datasets:
+              - sktime/base/**
+              - sktime/datasets/**
+            pyproject:
+              - pyproject.toml
```
Comment (yarnabrina) [inline .github/workflows/test.yml:210]: I needed it for testing extras can be installed or not, but you removed it. This can possibly be removed.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -186,6 +189,47 @@ jobs:
...
 
+  detect:
+    needs: test-nosoftdeps
+    name: detect changes
+    runs-on: ubuntu-latest
+    permissions:
+      pull-requests: read
+    outputs:
+      datasets: ${{ steps.filter.outputs.datasets }}
+      pyproject: ${{ steps.filter.outputs.pyproject }}
```
Comment (yarnabrina) [inline .github/workflows/test.yml:200]: Same as above. I don't think you need it any more.

**Review by yarnabrina — APPROVED** [review]
I really don't like few names in the old CI. For example, `test-nosoftdeps` calls `test_softdeps` and so on, and I have no idea what does that mean. Is it testing what happens in absence of soft dependencies, or is it testing what happens if all soft dependencies are present. But these are just my complaints, out of scope of this PR.

Added my non-blocking comments. Approved to merge.

**Comment by fkiraly** [comment]
> I really don't like few names in the old CI. For example, `test-nosoftdeps` calls `test_softdeps` and so on, and I have no idea what does that mean.

Well, so you understand my gripe with the CI element formerly known as `test_code_quality` and `validate_extras` then...

Perhaps we should sit down and think really well about good names, since now the two of us have consensus about that it's really not easy to find them.

> Is it testing what happens in absence of soft dependencies, or is it testing what happens if all soft dependencies are present.

But yes, that is exactly what it means! `test-nosoftdeps` tests what happens in the absence of soft dependencies, and `test_softdeps` tests what happens when they are present.

The condition/call afaik is there because if things fail without softdeps, we don't need to test with softdeps to know there is a failure (and the latter also eats more runtime/resource).

File: .github/workflows/test.yml (hunk 1 above)
Comment (fkiraly) [reply .github/workflows/test.yml:210]: hm, can you explain? I don't quite get your statement here.

Comment (yarnabrina) [reply .github/workflows/test.yml:210]: I don't think you are using "detection of change in pyproject.toml" anywhere after your modifications. So my guess is removing these two lines, plus outset creation to store result of that detection check is not necessary anymore.

Comment (fkiraly) [reply .github/workflows/test.yml:210]: Ah, I see. You are right.

May I suggest we leave it in for now, as this is not the final step of rework?
E.g., one may like to move some parts of the "detects"-s further down into that one step. Or, trigger extensive tests only if `pyproject` changes, etc.

Pending of course discussion of how we reduce the footprint of the "old tests".

**Comment by fkiraly** [comment]
merging since not blocking - let's continue improvements/changes in further PR.
